pkg/controller/instance/instance_controller.go

```diff
@@ -115,295 +76,201 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 	}
 
-	log.Printf("InstanceController: Received Reconcile request for instance \"%+v\"", request.Name)
+	return &activePlan{
+			Name:       activePlanStatus.Name,
+			Spec:       &planSpec,
+			PlanStatus: activePlanStatus,
+			Tasks:      ov.Spec.Tasks,
+			Templates:  ov.Spec.Templates,
+			params:     params,
+		}, &executionMetadata{
+			operatorVersionName: ov.Name,
+			operatorVersion:     ov.Spec.Version,
+			resourcesOwner:      instance,
+			operatorName:        ov.Spec.Operator.Name,
+			instanceNamespace:   instance.Namespace,
+			instanceName:        instance.Name,
+		}, nil
+}
 
-	// if this is new create and create and assign a planexecution and return
-	if isNewInstance(instance) {
-		err = createPlanAndUpdateReference(r.Client, r.recorder, r.scheme, "deploy", instance)
-		// err or not we return.  If err == nil the controller is done, else requeue
-		return reconcile.Result{}, err
+// handleError handles execution error by logging, updating the plan status and optionally publishing an event
+// specify eventReason as nil if you don't wish to publish a warning event
+// returns err if this err should be retried, nil otherwise
+func (r *Reconciler) handleError(err error, instance *kudov1alpha1.Instance) error {
+	log.Printf("InstanceController: %v", err)
+
+	// first update instance as we want to propagate errors also to the `Instance.Status.PlanStatus`
+	clientErr := r.Client.Update(context.TODO(), instance)
+	if clientErr != nil {
+		log.Printf("InstanceController: Error when updating instance state. %v", clientErr)
+		return clientErr
 	}
 
-	// Make sure all the required parameters in the operatorVersion are present
-	for _, param := range ov.Spec.Parameters {
-		if param.Required && param.Default == nil {
-			if _, ok := instance.Spec.Parameters[param.Name]; !ok {
-				r.recorder.Event(instance, "Warning", "MissingParameter", fmt.Sprintf("Missing parameter \"%v\" required by operatorversion \"%v\"", param.Name, ov.Name))
-			}
+	// determine if retry is necessary based on the error type
+	if exErr, ok := err.(*executionError); ok {
+		if exErr.eventName != nil {
+			r.Recorder.Event(instance, "Warning", kudo.StringValue(exErr.eventName), err.Error())
+		}
+
+		if exErr.fatal {
+			return nil // not retrying fatal error
 		}
 	}
 
-	// Defer call from above should apply the status changes to the object
-	return reconcile.Result{}, nil
+	// for code being processed on instance, we need to handle these errors as well
+	if iError, ok := err.(*kudov1alpha1.InstanceError); ok {
+		if iError.EventName != nil {
+			r.Recorder.Event(instance, "Warning", kudo.StringValue(iError.EventName), err.Error())
+		}
+	}
+	return err
+}
+
+// getInstance retrieves the instance by namespaced name
+// returns nil, nil when instance is not found (not found is not considered an error)
+func (r *Reconciler) getInstance(request ctrl.Request) (instance *kudov1alpha1.Instance, err error) {
```

Comment:
it looks like this should take a `ctx context.Context` as an argument.  This function is called once from `Reconcile` and each step creates it's own context (which works of course) but the point of context is to carry the context throughout the execution.

same with `getOperatorVersion`

Reply:
additionally, would it be helpful to create our own Context struct embedding conext.Context here?

Reply:
Also, to counter this, including myself, we should question ourselves every time we think we need a context somewhere because context typically ends up being _very_ viral in your codebase. I don't think it's an antipattern but should be very carefully considered as a whole before we do that.

Reply:
I am not sure, let's talk about this... right now we don't really use the context at all, we always use `context.TODO()` ... until we start doing something meaningful with it I don't see much advantage of having it as dependency as it's only bloating the number of parameters that method has... Having it as dependency and then always passing in context.TODO() for now that feels a bit meaningless